bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableIO.java

```diff
@@ -0,0 +1,1039 @@
+     * @param options The pipeline options.
+     * @return A list of sources split into groups.
+     */
+    @Override
+    public List<? extends BoundedSource<Result>> split(long desiredBundleSizeBytes,
+        PipelineOptions options) throws Exception {
+      // Update the desiredBundleSizeBytes in order to limit the number of splits to
+      // MAX_SPLIT_COUNT.  This is an extremely rough estimate for large scale jobs.  There is
+      // currently a hard limit on both the count of Sources as well as the sum of the sizes of
+      // serialized Sources.  This solution will not work for large workloads for cases where either
+      // the row key sizes are large, or the scan is large.
+      //
+      // TODO: Work on a more robust algorithm for splitting that works for more cases.
+      List<? extends BoundedSource<Result>> splits = getSplits(desiredBundleSizeBytes);
+      SOURCE_LOG.info("Creating {} splits.", splits.size());
+      SOURCE_LOG.debug("Created splits {}.", splits);
+      return splits;
+    }
+
+    /**
+     * Gets an estimated size based on data returned from {@link BigtableDataClient#sampleRowKeys}.
+     * The estimate will be high if a {@link Scan} is set on the {@link CloudBigtableScanConfiguration};
+     * in such cases, the estimate will not take the Scan into account, and will return a larger estimate
+     * than what the {@link CloudBigtableIO.Reader} will actually read.
+     *
+     * @param options The pipeline options.
+     * @return The estimated size of the data, in bytes.
+     */
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) throws IOException {
+      long totalEstimatedSizeBytes = 0;
+
+      byte[] scanStartKey = getConfiguration().getZeroCopyStartRow();
+      byte[] scanEndKey = getConfiguration().getZeroCopyStopRow();
+
+      byte[] startKey = HConstants.EMPTY_START_ROW;
+      long lastOffset = 0;
+      for (SampleRowKeysResponse response : getSampleRowKeys()) {
+        byte[] currentEndKey = response.getRowKey().toByteArray();
+        // Avoid empty regions.
+        if (Bytes.equals(startKey, currentEndKey) && startKey.length != 0) {
+          continue;
+        }
+        long offset = response.getOffsetBytes();
+        if (isWithinRange(scanStartKey, scanEndKey, startKey, currentEndKey)) {
+          totalEstimatedSizeBytes += (offset - lastOffset);
+        }
+        lastOffset = offset;
+        startKey = currentEndKey;
+      }
+      SOURCE_LOG.info("Estimated size in bytes: " + totalEstimatedSizeBytes);
+
+      return totalEstimatedSizeBytes;
+    }
+
+    @Override
+    public Coder<Result> getDefaultOutputCoder() {
+      // TODO Auto-generated method stub
+      return super.getDefaultOutputCoder();
+    }
+  }
+
+  /**
+   * A {@link BoundedSource} for a Cloud Bigtable {@link Table} with a start/stop key range, along
+   * with a potential filter via a {@link Scan}.
+   */
+  protected static class SourceWithKeys extends AbstractSource {
```

Comment:
The class name is a bit opaque. Maybe rename it to something that describes what its for.  Something like Shard or Split or SourceShard